### PR TITLE
I/6106: Introduced color pickers in the table and table cell property forms

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced color pickers in the table and table cell property forms. Introduced `config.table.tableProperties` and `config.table.tableCellProperties` configurations. Closes ckeditor/ckeditor5/issues/6106.

---

### Additional information

**Requires:**
* [ ] https://github.com/ckeditor/ckeditor5-theme-lark/pull/266 (styles for the picker, some styles refactoring)
* [x] https://github.com/ckeditor/ckeditor5-ui/pull/541 (added color grid utils that are now common for ckeditor5-table and ckeditor5-font)
* [x] https://github.com/ckeditor/ckeditor5-font/pull/59 (moved color grid utils to ckeditor5-ui)

**Some follow-ups**
* https://github.com/ckeditor/ckeditor5/issues/6237 (we need a single color palette configuration for all editor features, now what we have is a mess).
* https://github.com/ckeditor/ckeditor5/issues/6238 (let's have text ellipsis on all labels across the UI).
* https://github.com/ckeditor/ckeditor5/issues/6246 (get rid of dangling commas in the project)

image

### CODE REVIEW

**Comment by jodator** [comment]
Follow-ups:
- https://github.com/ckeditor/ckeditor5/issues/6239

File: src/tablecellproperties/ui/tablecellpropertiesview.js [hunk 1]
```diff
@@ -562,7 +595,7 @@ export default class TableCellPropertiesView extends View {
 
 		widthInput.set( {
 			label: t( 'Width' ),
-			class: 'ck-table-cell-properties-form__width',
+			class: 'ck-table-properties-form__width',
```
Comment (jodator) [inline src/tablecellproperties/ui/tablecellpropertiesview.js]: Why this change? It looks like an error. Did we extracted some common styles that I've missed?

File: src/tableproperties.js [hunk 2]
```diff
@@ -36,3 +36,45 @@ export default class TableProperties extends Plugin {
...
+/**
+ * A configuration of the table properties user interface (balloon). It allows to define:
+ *
+ * * the color palette for the table border color style field (`tableProperties.border.colors`),
+ * * the color palette for the table background style field (`tableProperties.backgroundColors`).
+ *
+ *		const tableConfig = {
+ *			tableProperties: {
+ *				border: {
+ *					colors: [
```
Comment (jodator) [inline src/tableproperties.js]: Same thing about config namespace.

File: src/ui/utils.js [hunk 3]
```diff
@@ -235,10 +236,215 @@ export function fillToolbar( { view, icons, toolbar, labels, propertyName, nameT
...
+		label: 'Blue'
+	},
+	{
+		color: 'hsl(270, 75%, 60%)',
+		label: 'Purple'
+	}
+];
+
+/**
+ * A function that helps creating labeled color inputs.
```
Comment (jodator) [inline src/ui/utils.js]:
```suggestion
 * Returns a creator for color input with a label.
```

File: tests/manual/tableproperties.js [hunk 4]
```diff
@@ -27,13 +27,12 @@ ClassicEditor
 		],
 		table: {
 			contentToolbar: [ 'tableColumn', 'tableRow', 'mergeTableCells', 'tableProperties', 'tableCellProperties' ],
-			tableToolbar: [ 'bold', 'italic' ]
-		}
+			tableToolbar: [ 'bold', 'italic' ],
```
Comment (jodator) [inline tests/manual/tableproperties.js]: dangling comma

File: tests/manual/tableproperties.js [hunk 5]
```diff
@@ -27,13 +27,12 @@ ClassicEditor
...
+		},
```
Comment (jodator) [inline tests/manual/tableproperties.js]: dangling comma

Comment (jodator) [reply tests/manual/tableproperties.js]: Check your IDE setting guys :D

File: tests/ui/colorinputview.js [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+		it( 'should set #isReadOnly', () => {
+			expect( view.isReadOnly ).to.be.false;
+		} );
+
+		it( 'should set #hasError', () => {
+			expect( view.hasError ).to.be.false;
+		} );
+
+		describe( 'dropdown', () => {
```
Comment (jodator) [inline tests/ui/colorinputview.js]:
```suggestion
		describe( '#_dropdownView', () => {
```

I'm torn between exposing this as public property - as testing on private fields/methods should be the last resort. OTOH this is that bad.

The only other solution I see here it to access `dropdownView` by `parent.chidlren` in `beforeEach` (similarly to `colorGridView`). But this is a minor thing and isn't by any means required. The minimal change here is to change `describe` description IMO.

File: tests/ui/colorinputview.js [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+				view.value = 'hsl(0, 0, 50%)';
+
+				expect( noColorPreview.classList.contains( 'ck-hidden' ) ).to.be.true;
+
+				view.value = '';
+
+				expect( noColorPreview.classList.contains( 'ck-hidden' ) ).to.be.false;
+			} );
+
+			it( 'should have the color grid', () => {
```
Comment (jodator) [inline tests/ui/colorinputview.js]: there's already a `color view` suite so those test doesn't belong here. Also there's `colorGridView`.

File: src/tablecellproperties/ui/tablecellpropertiesview.js (hunk 1 above)
Comment (oleq) [reply src/tablecellproperties/ui/tablecellpropertiesview.js]: Yeah, I noticed the duplication and I extracted them in this PR.

Comment (oleq) [reply src/tablecellproperties/ui/tablecellpropertiesview.js]: I went even further in https://github.com/ckeditor/ckeditor5-table/pull/244/commits/dc063334e91bad1d8b99a3817e1bc7caad639a5f.

File: tests/manual/tableproperties.js (hunk 5 above)
Comment (oleq) [reply tests/manual/tableproperties.js]: Can we configure ESLint to disallow it?

Comment (jodator) [reply tests/manual/tableproperties.js]: Would be awesome :D

Comment (Reinmar) [reply tests/manual/tableproperties.js]: My OCD would love this :P

File: src/tableproperties.js (hunk 2 above)
Comment (Reinmar) [reply src/tableproperties.js]: I left a comment in https://github.com/ckeditor/ckeditor5/issues/6106. I think it's an unnecessary complication.

File: src/tablecellproperties.js [hunk 8]
```diff
@@ -35,3 +35,45 @@ export default class TableCellProperties extends Plugin {
...
+
+/**
+ * A configuration of the table cell properties user interface (balloon). It allows to define:
+ *
+ * * the color palette for the cell border color style field (`tableCellProperties.border.colors`),
+ * * the color palette for the cell background style field (`tableCellProperties.backgroundColors`).
+ *
+ *		const tableConfig = {
+ *			tableCellProperties: {
+ *				border: {
```
[the remark this answers is not in the document]
Comment (Reinmar) [reply src/tablecellproperties.js]: Done.

File: src/ui/utils.js [hunk 9]
```diff
@@ -235,10 +236,215 @@ export function fillToolbar( { view, icons, toolbar, labels, propertyName, nameT
...
+ *			{
+ *				color: 'hsl(270, 75%, 60%)',
+ *				label: 'Purple'
+ *			}
+ *		];
+ */
+export const defaultColors = [
+	{
+		color: 'hsl(0, 0%, 0%)',
+		label: 'Black'
```
Comment (Reinmar) [inline src/ui/utils.js]: Just to be sure –  those get translated?

**Comment by Reinmar** [comment]
One more followup https://github.com/ckeditor/ckeditor5/issues/6241.

File: src/ui/utils.js (hunk 9 above)
Comment (panr) [reply src/ui/utils.js]: Yes, config (default and this set by user) passes through https://github.com/ckeditor/ckeditor5-table/pull/244/files#diff-9b951d2182502e8a107b93340ea85135R144

File: tests/manual/tableproperties.js (hunk 5 above)
Comment (oleq) [reply tests/manual/tableproperties.js]: [There you go.](https://github.com/ckeditor/ckeditor5/issues/6246)

File: tests/ui/colorinputview.js (hunk 6 above)
Comment (oleq) [reply tests/ui/colorinputview.js]: Um, that's why the `#_dropdownView` is `@protected`. Besides, this protected prop must exist anyway because it's used across methods in the component (they open and close it).

Comment (oleq) [reply tests/ui/colorinputview.js]: Just FYI "dropdown" here is a natural language, not the name of the property in the code.

Comment (jodator) [reply tests/ui/colorinputview.js]: Sure, can be left as is.

**Comment by Reinmar** [comment]
cc @ckeditor/qa-team

File: src/tablecellproperties.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (oleq) [reply src/tablecellproperties.js]: It's been fixed in https://github.com/ckeditor/ckeditor5-table/pull/244/commits/dfebee20504d4a6fa898ecb46f435a2157b507c8.

File: src/tableproperties.js (hunk 2 above)
Comment (oleq) [reply src/tableproperties.js]: See https://github.com/ckeditor/ckeditor5-table/pull/244#discussion_r378725172

**Comment by Mgsy** [comment]
Looks good 👍
